include/downloader_pool.h

```diff
@@ -0,0 +1,248 @@
+		// Because of language limitations I cannot put this in the end of class
+		// For better clarity using simple struct instead of std::tuple
+		struct WorkingTuple
+		{
+			Downloaders::CHTTPDownloader<SocketClass> m_downloader;
+			CURI m_URIToDownload;
+			std::thread m_runningThread;
+			FDownloadCallback m_fCallback;
+
+			WorkingTuple(const CURI &newURI) : m_URIToDownload{newURI}
+			{
+
+			}
+		};
+	public:
+		CConcurrentDownloader(unsigned uMaxCountOfThreads) noexcept : 
+			m_runningDownloadsLock{new(std::nothrow) std::mutex()},
+			m_uriQueueLock{new(std::nothrow) std::mutex()},
+			m_uMaxCountOfThreads{uMaxCountOfThreads},
+			m_bShouldStop{new(std::nothrow) bool(false)},
+			m_bCanAddNewTasks{new(std::nothrow) bool(true)}
+		{
+
+		}
+		CConcurrentDownloader() noexcept :
+			m_runningDownloadsLock{new(std::nothrow) std::mutex()},
+			m_uriQueueLock{new(std::nothrow) std::mutex()},
+			m_uMaxCountOfThreads{std::thread::hardware_concurrency()},
+			m_bShouldStop{new(std::nothrow) bool(false)},
+			m_bCanAddNewTasks{new(std::nothrow) bool(true)}
+		{
+
+		}
+		~CConcurrentDownloader() noexcept
+		{
+			// Clearing queue of URIs put for download
+			std::queue<std::pair<size_t, WorkingTuple>>().swap(m_urisToDownload);
+
+			// Setting stop flag if we can do that
+			if(m_bShouldStop && m_bCanAddNewTasks)
+			{
+				*m_bShouldStop = true;
+				*m_bCanAddNewTasks = false;
+			}
+
+			// Joining all threads if we can
+			if(m_runningDownloadsLock)
+			{
+				m_runningDownloadsLock->lock();
+				for(auto &dataPair : m_runningDownloads)
+				{
+					if(dataPair.second.m_runningThread.joinable())
+					{
+						dataPair.second.m_runningThread.join();
+					}
+				}
+				m_runningDownloadsLock->unlock();
+			}
+
+			DealocateResources();
+		}
+		
+		// We cannot copy threads
+		CConcurrentDownloader(const CConcurrentDownloader &cCopyOfDownloader) = delete;
+		CConcurrentDownloader &operator =(const CConcurrentDownloader &cCopyOfDownloader) = delete;
+		// We can change their owner
+		CConcurrentDownloader(CConcurrentDownloader &&downloaderToMove) noexcept
+		{
+			MoveData(downloaderToMove);
+		}
+		CConcurrentDownloader &operator =(CConcurrentDownloader &&downloaderToMove) noexcept
+		{
+			if(this != &downloaderToMove)
+			{
+				MoveData(std::move(downloaderToMove));
+			}
+
+			return *this;
+		}
+
+		/// Function to add new URI for download
+		void AddNewTask(const CURI &cURIToDownload, FDownloadCallback fCallback) noexcept
+		{
+			// We can't add new tasks if we already stoped processing
+			if(*m_bShouldStop || !*m_bCanAddNewTasks)
+			{
+				return;
+			}
+			std::unique_lock<std::mutex> urisLock{*m_uriQueueLock};
+
+			WorkingTuple workerData{cURIToDownload};
+			workerData.m_fCallback = fCallback;
+
+			m_urisToDownload.push(std::make_pair(m_nCurrentID, std::move(workerData)));
+			++m_nCurrentID;
+			urisLock.unlock();
+
+			LoadQueueIntoDownloads();
+		}
+
+		void JoinTasksCompletion() noexcept
+		{
+			*m_bCanAddNewTasks = false;
```

Comment:
Should this part also be synchronized?

Reply:
But bools are only set in destructor and  JoinTasks functions